module/epochs/qc_voter.go

```diff
@@ -121,42 +123,53 @@ func (voter *RootQCVoter) Vote(ctx context.Context, epoch protocol.Epoch) error
 		}
 
 		// check whether we've already voted, if we have we can exit early
-		voted, err := voter.qcContractClient().Voted(ctx)
+		voted, err := qcContractClient.Voted(ctx)
 		if err != nil {
 			log.Error().Err(err).Msg("could not check vote status")
 			return retry.RetryableError(err)
 		} else if voted {
 			log.Info().Msg("already voted - exiting QC vote process...")
+			// update our last successful client index for future calls
+			voter.lastSuccessfulClientIndex = clientIndex
 			return nil
 		}
 
 		// submit the vote - this call will block until the transaction has
 		// either succeeded or we are able to retry
 		log.Info().Msg("submitting vote...")
-		err = voter.qcContractClient().SubmitVote(ctx, vote)
+		err = qcContractClient.SubmitVote(ctx, vote)
 		if err != nil {
 			log.Error().Err(err).Msg("could not submit vote - retrying...")
 			return retry.RetryableError(err)
 		}
 
 		log.Info().Msg("successfully submitted vote - exiting QC vote process...")
 
+		// update our last successful client index for future calls
+		voter.lastSuccessfulClientIndex = clientIndex
 		return nil
 	})
 
 	return err
 }
 
-func (voter *RootQCVoter) qcContractClient() module.QCContractClient {
-	return voter.qcContractClients[voter.activeQCContractClient]
-}
+// updateContractClient will return the last successful client index by default for all initial operations or else
+// it will return the appropriate client index with respect to last successful and number of client.
+func (voter *RootQCVoter) updateContractClient(clientIndex int, init bool) (int, module.QCContractClient) {
+	// if initial call return lastSuccessfulClientIndex will either be the index of the last successful client or 0
+	if init {
+		return voter.lastSuccessfulClientIndex, voter.qcContractClients[voter.lastSuccessfulClientIndex]
+	}
 
-func (voter *RootQCVoter) updateActiveQcContractClient() {
-	// if we have reached the end of our array start from beginning
-	if voter.activeQCContractClient == len(voter.qcContractClients)-1 {
-		voter.activeQCContractClient = 0
-		return
+	if clientIndex == voter.lastSuccessfulClientIndex {
```

Comment:
To me it would be simpler / less code to use `voter. lastSuccessfulClientIndex` directly instead of the intermediary variable. That way we can encapsulate mutex-protected reads/writes of that variable in the `updateContractClient` and `getInitialContractClient` methods too. 

Not a blocker though.

Reply:
It's a little more verbose but more explicit